### PR TITLE
CSV Uploader alpha for bulk import - Invoices [ch24191]

### PR DESCRIPTION
**Pull request by macharmi**
I took the CSV definition from [our official documentation](https://help.chartmogul.com/hc/en-us/articles/360013881659-Import-billing-system-data-using-CSV-upload)

We should be able to create ChartMogul::CSV::Invoice.new with keywords:

```
invoice = ChartMogul::CSV::Invoice.new(customer_external_id: 'customer_id', external_id: 'invoice_id', date: Time.new(2020, 8, 24, 8, 22, 15), due_date: Time.new(2020, 8, 31, 23, 59, 59), currency: 'USD')
```

We have to methods:
- `to_csv`

```
2.7.0 :060 > invoice.to_csv
 => #<struct customer_external_id="customer_id", external_id="invoice_id", date=2020-08-24 08:22:15 +0200, due_date=2020-08-31 23:59:59 +0200, currency="USD"> 
2.7.0 :061 > invoice.to_csv.to_a
 => ["customer_id", "invoice_id", 2020-08-24 08:22:15 +0200, 2020-08-31 23:59:59 +0200, "USD"] 
```
- `csv_file_headers`

```
2.7.0 :064 > ChartMogul::CSV::Invoice.csv_file_headers
 => #<struct customer_external_id="Customer external ID", external_id="Invoice external ID", date="Invoiced date", due_date="Due date", currency="Currency"> 
2.7.0 :065 > ChartMogul::CSV::Invoice.csv_file_headers.to_a
 => ["Customer external ID", "Invoice external ID", "Invoiced date", "Due date", "Currency"] 
```

### CODE REVIEW

File: lib/chartmogul/csv/csv_header.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ChartMogul
+  module CSV
+    module CSVHeader
+      def csv_file_headers
+        Struct.new(*fields).new(*headers)
```
Comment (pkopac) [inline lib/chartmogul/csv/csv_header.rb:7]: This creates a different anonymous class each time you call this :sweat: please move into a constant named as a class, eg: `Headers = Struct.new(*fields)`

File: lib/chartmogul/csv/csv_object.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module ChartMogul
+  module CSV
+    module CSVObject
+      def to_csv
+        data = fields.map { |e| send(e) }
+        Struct.new(*fields).new(*data)
```
[the remark this answers is not in the document]
Comment (pkopac) [reply lib/chartmogul/csv/csv_object.rb:8]: This line creates a new anonymous class on each call :sweat: 
I would suggest the class itself is a `ChartMogul::CSV::Invoice.is_a? Struct == true`
```
Invoice = Struct.new(...) do
  methods if necessary
end
```
then as you pointed out before, you don't need any additional method:
```
[1] pry(main)> A = Struct.new(:a, :b)
=> A
[2] pry(main)> a = [A.new(1, 2), A.new(3, 4)]
=> [#<struct A a=1, b=2>, #<struct A a=3, b=4>]
[3] pry(main)> CSV.generate { |csv| csv << a[0]; csv << a[1] }
=> "1,2\n3,4\n"
```

**Review by pkopac — CHANGES_REQUESTED** [review]

**Comment by pkopac** [comment]
@macharmi regarding release - we should create a feature branch `alpha`, merge these PRs there and release those versions under `2.0.0-alpha.1`, `2.0.0-alpha.2` etc., so that standard users don't even get these classes. When we finalize the implementation, we can release it as 2.0.0 CSV upload support. WDYT?

File: lib/chartmogul/csv/csv_object.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (macharmi) [reply lib/chartmogul/csv/csv_object.rb]: I would like to continue to init `ChartMogul::CSV::Invoice` by a `Hash`. Relying on args order is just like a hell. So, I would like to continue to wrap the Struct in a class.

File: lib/chartmogul/csv/invoice.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ChartMogul
+  module CSV
+    class Invoice < ChartMogul::Object
```
Comment (pkopac) [inline lib/chartmogul/csv/invoice.rb:5]: I don't think this needs to be `ChartMogul::Object`.
```
Invoice = Struct.new(*FIELDS) do
  include/extend ...
```
just works. You can access and write to the fields likewise, so in fact this `writeable_attr` is not necessary. Please simplify.

**Review by pkopac — CHANGES_REQUESTED** [review]

File: lib/chartmogul/csv/invoice.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ChartMogul
+  module CSV
```
Comment (pkopac) [inline lib/chartmogul/csv/invoice.rb:4]: documenting this [link](https://help.chartmogul.com/hc/en-us/articles/360013881659-Import-billing-system-data-using-CSV-upload#h_e4e19bd1-38a1-432d-a8f0-9f4b52bee0d7) might be a good idea around here.

**Review by pkopac — APPROVED** [review]

**Comment by pkopac** [comment]
@macharmi oh and you need to add alpha branch here: https://github.com/chartmogul/chartmogul-ruby/blob/alpha/.travis.yml#L22 so that the build checks are run :)

**Pull request merged**